tools/encryption/pkg/encryption-helper.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	r := keyset.NewJSONReader(resp.Body)
+	ks, err := r.Read()
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	return *ks, nil
+}
+
+// Generates a new AES-GCM key.
+func generateNewAesGcmKey(km registry.KeyManager) ([]byte, error) {
+	p, err := proto.Marshal(&gcmpb.AesGcmKeyFormat{KeySize: aesGcmKeySize})
+	if err != nil {
+		return nil, err
+	}
+	m, err := km.NewKey(p)
+	if err != nil {
+		return nil, err
+	}
+	return proto.Marshal(m)
+}
+
+// Creates an AES-GCM Keyset using the input key.
+func createAesGcmKeyset(key []byte) tinkpb.Keyset {
+	keyData := tinkpb.KeyData{
+		KeyMaterialType: tinkpb.KeyData_SYMMETRIC,
+		TypeUrl:         aesGcmKeyUrl,
+		Value:           key,
+	}
+	keys := []*tinkpb.Keyset_Key{
+		&tinkpb.Keyset_Key{
+			KeyData:          &keyData,
+			Status:           tinkpb.KeyStatusType_ENABLED,
+			KeyId:            1,
+			OutputPrefixType: tinkpb.OutputPrefixType_TINK,
+		},
+	}
+	return tinkpb.Keyset{
+		PrimaryKeyId: 1,
+		Key:          keys,
+	}
+}
+
+// Retrieves all encrypted content sections from the parsed HTML tree.
+func getAllEncryptedSections(parsedHtml *html.Node) []*html.Node {
+	for n := parsedHtml.FirstChild; n != nil; n = n.NextSibling {
+		if (n.DataAtom == atom.Html) && (len(n.Attr) != 0) {
+			for bn := n.FirstChild; bn != nil; bn = bn.NextSibling {
+				if bn.DataAtom == atom.Body {
+					return getEncryptedSectionsDfs(bn)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// Searches for all <section subscriptions-section="content" encrypted> nodes and returns them.
+func getEncryptedSectionsDfs(bodyNode *html.Node) []*html.Node {
+	var encryptedSections []*html.Node
+	var stack []*html.Node
+	var n *html.Node
+	stack = append(stack, bodyNode)
+	for {
+		if len(stack) == 0 {
+			break
+		}
+		n, stack = stack[len(stack)-1], stack[:len(stack)-1]
+		if n.Type == html.ElementNode && n.DataAtom == atom.Section {
+			var contentSubSection bool = false
+			var encrypted bool = false
+			for _, a := range n.Attr {
+				if a.Key == "subscriptions-section" && a.Val == "content" {
+					contentSubSection = true
+				} else if a.Key == "encrypted" {
+					encrypted = true
+				}
+			}
+			if contentSubSection && encrypted {
+				encryptedSections = append(encryptedSections, n)
+			}
+		}
+		for cn := n.FirstChild; cn != nil; cn = cn.NextSibling {
+			stack = append(stack, cn)
+		}
+	}
+	return encryptedSections
+}
+
+// Encrypts the content inside of the input "encryptedSections" nodes.
+func encryptAllSections(parsedHtml *html.Node, encryptedSections []*html.Node, kh *keyset.Handle) error {
+	cipher, err := aead.New(kh)
+	if err != nil {
+		return err
+	}
+	for _, node := range encryptedSections {
+		var content []string
+		for c := node.FirstChild; c != nil; c = c.NextSibling {
+			content = append(content, renderNode(c, true))
+			node.RemoveChild(c)
+		}
+		encContent, err := cipher.Encrypt([]byte(strings.Join(content, "")), nil)
```

Comment:
My old question: can you confirm that the content joined here is all in UTF-8?

Reply:
Added a call to utf8.Valid before encrypting.

Reply:
This would seem to be a wrong place to check for valid UTF8 since you've already processed the input stream and parsed the HTML which inherently has to decode a byte stream into a char stream. Otherwise it'd not be able to scan characters such as `<` and `>`.

Reply:
The only thing that really matters is the original stream that goes into html.Parse. I think we just need to make sure our tests (manual and/or otherwise) have characters from non-ASCII set.

Reply:
Acknowledged.